### ISSUE TITLE
Implement ArrayHelper::replace and ArrayHelper::replaceMulti

### ISSUE THREAD

**Issue by mikehaertl** [yiisoft/yii2#9580]
I suggest to add the following two methods to `ArrayHelper`:

```php
    /**
     * Replace all occurences of a value in an array
     *
     * @param array $haystack the array to replace values in
     * @param mixed $needle the value to replace
     * @param mixed $replacement the replacement value
     * @return array the array with the values replace
     */
    static function replace($haystack, $needle, $replacement)
    {
        return array_map(function ($v) use ($needle, $replacement) {
            return $v===$needle ? $replacement : $v;
        }, $haystack);
    }

    /**
     * Replace all occurences of several values in an array
     *
     * @param array $haystack the array to replace values in
     * @param array $replacements replacement values indexed by the values to replace
     * @return array the array with the values replace
     */
    static function replaceMulti($haystack, $replacements)
    {
        return array_map(function ($v) use ($replacements) {
            return isset($replacements[$v]) ? $replacements[$v] : $v;
        }, $haystack);
    }
```

**Comment by klimov-paul** [comment]
What whould be a practical use case?

**Comment by mikehaertl** [comment]
Well, we have `ArrayHelper::remove()` so the use cases would probably be quite similar. It's really just a convenience method that may come in handy in different situtations. I needed this in a current project and had to extend `ArrayHelper` to add it. So I wondered, why not include it in the core.

**Comment by mikehaertl** [comment]
Just for the first time found out, that `ArrayHelper::remove()` doesn't actually do, what I thought it would (searching for a *value* and removing it, if found). Nevertheless, my suggested addition might still be useful.

**Comment by dynasource** [comment]
Can you tell me in what kind of scenario, one would use this function?

Btw, there is a bug in your code: $replacment

**Comment by mikehaertl** [comment]
In any situation, where you might have to replace an array value. My example is very application specific, so it doesn't make sense to bring it up here. It's not such an unusual use case, as you can see from the views and upvotes here: 

http://stackoverflow.com/questions/8668826/search-and-replace-value-in-php-array

I don't mind, if it's not included. But you should keep in mind, just because you can't think of a use case for you *now* doesn't mean, that there *is* no valid use case. There certainly is.

**Comment by dynasource** [comment]
> you should keep in mind, just because you can't think of a use case

there is a usecase for everything, resulting into an ArrayHelper of multiple MB's. The trick is to put only the functions inside that are used most so the framework can be kept small, lean and mean.

Perhabs your usecase is something very common but without an practical example, its hard to understand if this is something which is going to be used.

What would be good in this case is a general view on Arrays and the practical value of your functions in particular for projects.

**Comment by samdark** [comment]
Would be great to get a use case. No doubt it could be useful but I can't think of any use case currently which is a bad sign.

**Comment by mikehaertl** [comment]
I think, this is such a generic utility method, that it's hard to tell, where exactly you need it. Thing is: You'll know, when you need it and it isn't there.

In my case i had to create several different CSV product feeds from a shop DB. The classes layout makes it easy to create such an export:

```
/**
 * Creates a product feed for geizhals.de
 */
class Geizhals extends BaseFeed
{
    /**
     * @inheritdoc
     */
    public function getSelectColumns()
    {
        return [
            'Bezeichnung' => 'pt.name',
            'Hersteller' => 'm.name',
            'HerstNr' => 'p.mpn',
            [...]
        ];
    }

    /**
     * @inheritdoc
     */
    public function getFieldNames()
    {
        return ArrayHelper::replaceMulti(parent::getFieldNames(), [
            'HerstNr' => 'Herst.Nr.',
        ]);
    }
}
```

The return value of `getSelectColumns()` is an array that is fed into a query ($query->select($this->getSelectColumns())`. So the keys must not contain any dots (`.`).

The same method is used to get the field names for line 1 of the CSV file - order is important! The implementation of `BaseFeed::getFieldNames()` thus is simply:

```
    public function getFieldNames()
    {
        return array_keys($this->getSelectColumns());
    }
```

You see, that for `Geizhals` (a german price search engine) I had to override this method to replace `HerstNr` for t CSV field label back to `Herst.Nr.`.

You've been warned: The use case is very specific. ;)

**Comment by df2** [comment]
I'm not sure that as the use case evolves, so won't the complexity of `replace()` or `replaceMulti()` to the point that you're rewriting them yourself again anyway.

**Comment by mikehaertl** [comment]
Not sure, what you try to say. The use case is extremely generic. You have an array:

```
[
  0 => 'value1',
  1 => 'value2',
  2 => 'value3',
]
```

You now want to replace `value2` with `newvalue` without changing the array order. That's all.

**Comment by df2** [comment]
Could be named a bit clearer, like: `replaceValue()` and `replaceValueMulti()`

**Comment by mikehaertl** [comment]
Disagree. An array consists of values, so by default any function that modifies an array operates on the values. Same for all other native PHP methods. It's `array_map` not `array_map_values`, `sort`, not `sort_values` etc.

**Comment by df2** [comment]
:-D I see a use case!
```php
function listFriends($friends, $observer)
{
    if ($observer === 'wife')
    {
        $friends = ArrayHelper::replaceMulti($friends, [
            'Girlfriend1' => 'Cousin1',
            'Girlfriend2' => 'Cousin2',
            'Girlfriend3' => 'Cousin3',
        ]);
    }
    
    return $friends;
}
```

**Comment by samdark** [comment]
@df2 :)

**Comment by samdark** [comment]
@yiisoft/core-developers opinions?

**Comment by klimov-paul** [comment]
I can't see any practical benefit. Of course some usage can be found for anything, but in all my practice as a developer I have never met a task, which may require such feature.

Also there is a native PHP function [array-replace](http://php.net/manual/en/function.array-replace.php), which does something different.

**Comment by mikehaertl** [comment]
> ... but in all my practice as a developer I have never met a task, which may require such feature.

This is not really an argument, because you probably also never needed `array_product()` or similar - still they do exist, because in some programs you need such a thing. And if you start to think in functional programming patterns, powerful array helpers become a must.

I find replacing an array value to be such an elementary and simple thing, that i really was surprised that PHP doesn't offer a native function for it. I mean, you can do almost anything with Arrays. This is one of the few execptions. (Replacing a key and keeping the order would be another one).

But I can understand if it's not included to keep the core lightweight. So up to you.

**Comment by samdark** [comment]
Same as @klimov-paul, I've never needed something like that. I'd wait a bit for more requests of such functionality from community members.

**Comment by petrabarus** [comment]
I think while `\yii\helpers\ArrayHelper` can provide a lot of help in the application, I don't think we should put a lot of functionalities that aren't used in the framework components. I prefer to keep the size of the framework and use 3rd party library for array manipulations like this.

But that being said, I haven't found a really complete Array manipulation library for PHP.

**Comment by klimov-paul** [comment]
Resolved by #11899
